### PR TITLE
Feature/mock exchange

### PR DESCRIPTION
**Pull request by jgimeno**
Creating a backend that mocks requests to several exchanges, right now it only suppor exrates.

Important changes:
- [x] Something worth noting for reviewers.

Mandatory:
- [x] Added documentation.

Closes #
https://github.com/kowala-tech/kcoin/issues/726

### CODE REVIEW

File: mock-exchange/exchange/exrates/transformer.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
+package exrates
+
+import (
+	"encoding/json"
+
+	"github.com/kowala-tech/kcoin/mock-exchange/app"
+)
+
+type Transformer struct {
```
Comment (yourheropaul) [inline mock-exchange/exchange/exrates/transformer.go:9]: Should this not be an interface defined somewhere else?

```
type Transformer interface {
    Transform(request app.Request) (string, error)
}
```

and then there'd be an implementation for each exchange?

**Review by yourheropaul — CHANGES_REQUESTED** [review]

File: mock-exchange/exchange/exrates/transformer.go (hunk 1 above)
Comment (jgimeno) [reply mock-exchange/exchange/exrates/transformer.go:9]: Yes, you are right. I dont know why in the end this is here, I think I defined first in exchange folder. I am gonna returns this there, maybe I changed it when I got a cycle import problem.

Comment (jgimeno) [reply mock-exchange/exchange/exrates/transformer.go:9]: Ah wait! Check in exchange/

```package exchange

import "github.com/kowala-tech/kcoin/mock-exchange/app"

type Transformer interface {
	Transform(request app.Request) (string, error)
}
```

Comment (jgimeno) [reply mock-exchange/exchange/exrates/transformer.go:9]: Exactly what you mentioned!

File: e2e/features/mock_exchange.feature [hunk 2]
```diff
@@ -0,0 +1,19 @@
+Feature: Using the mocked exchange backend
+  As a user I want to use the backend
+  To perform fake controlled requests
+
+  Background:
+    Given the network is running
```
Comment (acroca) [inline e2e/features/mock_exchange.feature:6]: It seems this e2e test is just testing the mocked exchange. Do we really need a running network?
Also, do we really need to use e2e tests for this? I'd just use a standard go test instead.

Comment (JekaMas) [reply e2e/features/mock_exchange.feature:6]: Agree. A pack of unit tests seems more suitable for this.

Comment (acroca) [reply e2e/features/mock_exchange.feature:6]: Unless we plan to add e2e tests that really test how this exchange affect the network. Is it the plan @jgimeno ?

Comment (yourheropaul) [reply e2e/features/mock_exchange.feature:6]: We'll need e2e tests for the oracle client, which will use the mock exchange as data sources. So yes, there will eventually be full-network tests

Comment (jgimeno) [reply e2e/features/mock_exchange.feature:6]: yep, this is just a test as an example on how to use it later for the oracle.

**Review by yourheropaul — APPROVED** [review]

File: e2e/features/mock_exchange.feature (hunk 2 above)
Comment (jmatosp) [reply e2e/features/mock_exchange.feature:6]: yeah, I agree, don't think we need e2e test on a mock

File: mock-exchange/Gopkg.toml [hunk 3]
```diff
@@ -0,0 +1,54 @@
+# Gopkg.toml example
```
Comment (jmatosp) [inline mock-exchange/Gopkg.toml]: delete the provided example

File: mock-exchange/app/request.go [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package app
+
+type Request struct {
+	Sell []RateValue `json:"sell"`
+	Buy  []RateValue `json:"buy"`
+}
+
+type RateValue struct {
+	Amount float64
```
Comment (jmatosp) [inline mock-exchange/app/request.go:9]: nothing good will come out of a float. are you sure about using float?

File: mock-exchange/main.go [hunk 5]
```diff
@@ -0,0 +1,9 @@
+package main
+
+import (
+	"github.com/kowala-tech/kcoin/mock-exchange/cmd"
+)
+
+func main() {
+	cmd.Execute()
```
Comment (jmatosp) [inline mock-exchange/main.go]: magic!

don't think we should use this approach, use the standard cobra example. 
this example you used, is using a global variable

File: mock-exchange/server/handler.go [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+	Error string `json:"error"`
+}
+
+type FetchDataHandler struct {
+	cache *cache.Cache
+}
+
+func (h *FetchDataHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if r.Method != http.MethodPost {
+		logrus.WithFields(logrus.Fields{
```
Comment (jmatosp) [inline mock-exchange/server/handler.go]: you can shorten this to using this line at the beginning of the method and then just use the `logger` reference

```
logger := logrus.WithFields(logrus.Fields{
			"call":    "fetch",
			"method": r.Method.
		})

...
logger.WithField("request", request).Warn("invalid method called")
```

File: mock-exchange/server/handler.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+		w.WriteHeader(http.StatusMethodNotAllowed)
+		return
+	}
+
+	req, ok := h.cache.Get(app.CacheRequestKey)
+	if !ok {
+
+		jsonErrResp, err := json.Marshal(ErrorResponse{Error: "Please, call fetch before."})
+		if err != nil {
+			logrus.WithFields(logrus.Fields{
```
Comment (jmatosp) [inline mock-exchange/server/handler.go]: `logger.WithError(err).Warn("error creating error response")`

File: mock-exchange/server/handler.go [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+		logrus.WithFields(logrus.Fields{
+			"call": "getRate",
+		}).Info("msg", "get rates api method was called before fetching mocked info")
+		return
+	}
+
+	tResp, err := h.transformer.Transform(req.(app.Request))
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		jsonErrResp, err := json.Marshal(ErrorResponse{Error: "There was a problem getting data from request."})
```
Comment (jmatosp) [inline mock-exchange/server/handler.go]: maybe this can help reducing some of the boiler place
```
JSON(w, ErrorResponse{Error: "There was a problem getting data from request."}, http.StatusBadRequest)

// JSON prepares a response write to be written with all the headers needed
func JSON(output http.ResponseWriter, val interface{}, code int) {
	var content []byte
	var err error

	output.Header().Set("Content-Type", "application/json")
	output.WriteHeader(code)

	if val != nil {
		content, err = json.Marshal(val)
		output.Header().Set("Content-Length", strconv.Itoa(len(content)))
		if err != nil {
			http.Error(output, "", http.StatusInternalServerError)
			return
		}
		output.Write(content)
	}
}
```

File: e2e/features/mock_exchange.feature (hunk 2 above)
Comment (jgimeno) [reply e2e/features/mock_exchange.feature:6]: But this e2e is just created so in the future it is easy to launch e2e tests using the mock server. What I can do is remove the feature files, or comment. @jmatosp @yourheropaul

File: mock-exchange/app/request.go (hunk 4 above)
Comment (jgimeno) [reply mock-exchange/app/request.go:9]: @jmatosp for now there is no operations between floats, it is just a way to send floating point numbers, that then are printed back exactly the same. I dont think we will face tipycal float errors, what do you think?

Comment (jmatosp) [reply mock-exchange/app/request.go:9]: just pointing out that if we do any calculations we might go wrong, but then again this isn't the real thing :+1:

File: e2e/features/mock_exchange.feature (hunk 2 above)
Comment (yourheropaul) [reply e2e/features/mock_exchange.feature:6]: Maybe comment that line out for now so that the tests are quicker?
